spanner/spannertest/db_query.go

```diff
@@ -587,6 +585,167 @@ func (d *database) evalSelect(sel spansql.Select, params queryParams) (ri rowIte
+		}
+		t.mu.Lock()
+		ti := &tableIter{t: t}
+		if sf.Alias != "" {
+			ti.alias = sf.Alias
+		} else {
+			// There is an implicit alias using the table name.
+			// https://cloud.google.com/spanner/docs/query-syntax#implicit_aliases
+			ti.alias = sf.Table
+		}
+		ec.cols = ti.Cols()
+		return ec, ti, t.mu.Unlock, nil
+	case spansql.SelectFromJoin:
+		// TODO: Avoid the toRawIter calls here by rethinking how locking works throughout evalSelect,
+		// then doing the RHS recursive evalSelectFrom in joinIter.Next on demand.
+
+		lhsEC, lhs, unlock, err := d.evalSelectFrom(ec, sf.LHS)
+		if err != nil {
+			return ec, nil, nil, err
+		}
+		lhsRaw, err := toRawIter(lhs)
+		unlock()
+		if err != nil {
+			return ec, nil, nil, err
+		}
+
+		rhsEC, rhs, unlock, err := d.evalSelectFrom(ec, sf.RHS)
+		if err != nil {
+			return ec, nil, nil, err
+		}
+		rhsRaw, err := toRawIter(rhs)
+		unlock()
+		if err != nil {
+			return ec, nil, nil, err
+		}
+
+		// Construct a merged evalContext.
+		// TODO: Remove ambiguous names here? Or catch them when evaluated?
+		ec.cols = append(append([]colInfo(nil), lhsEC.cols...), rhsEC.cols...) // force a copy
+		ec.row = make(row, len(ec.cols))                                       // row is used as scratch space in joinIter.Next.
+		// TODO: aliases might need work?
+
+		return ec, &joinIter{
+			sfj: sf,
+			ec:  ec,
+
+			lhs:     lhsRaw,
+			rhsOrig: rhsRaw,
+		}, func() {}, nil
+	}
+}
+
+type joinIter struct {
+	sfj     spansql.SelectFromJoin // TODO: or less of this?
+	ec      evalContext            // combined context
+	scratch row                    // used for synthetic lhs+rhs row
+
+	// lhs is scanned (consumed), but rhs is cloned for each lhs row.
+	lhs, rhsOrig *rawIter
+
+	lhsRow row      // current row from lhs, or nil if it is time to advance
+	rhs    *rawIter // current clone of rhs
+	any    bool     // true if any rhs rows have matched lhsRow
+}
+
+func (ji *joinIter) Cols() []colInfo { return ji.ec.cols }
+
+func (ji *joinIter) nextLeft() error {
+	var err error
+	ji.lhsRow, err = ji.lhs.Next()
+	if err != nil {
+		return err
+	}
+	ji.rhs = ji.rhsOrig.clone()
+	ji.any = false
+	return nil
+}
+
+func (ji *joinIter) Next() (row, error) {
+	// TODO: More join types.
+	if ji.sfj.Type != spansql.LeftJoin {
+		return nil, fmt.Errorf("TODO: can't yet evaluate join of type %v in %s", ji.sfj.Type, ji.sfj.SQL())
+	}
+
+	/*
+		The result of a LEFT OUTER JOIN (or simply LEFT JOIN) for two
+		from_items always retains all rows of the left from_item in the
+		JOIN clause, even if no rows in the right from_item satisfy the
+		join predicate.
+
+		LEFT indicates that all rows from the left from_item are
+		returned; if a given row from the left from_item does not join
+		to any row in the right from_item, the row will return with
+		NULLs for all columns from the right from_item. Rows from the
+		right from_item that do not join to any row in the left
+		from_item are discarded.
+	*/
+	if ji.lhsRow == nil {
```

Comment:
nit: This causes the parser to accept anything in the ON condition if the right hand side is empty. In other words:

`SELECT * FROM JoinA AS A LEFT OUTER JOIN JoinB AS B ON whatever = other_whatever`

is accepted if `JoinB` is an empty table.

Reply:
Isn't that intended? That's the "even if no rows in the right from_item satisfy the join predicate" implies.

Anyway, I'll do some experiments and tweak this when adding some more join types.

Reply:
No, what I mean is that the ON clause can contain any invalid / unknown field names you want (and maybe even any jibberish you want, I didn't really look into that). Which means that the query will work fine as long as the `JoinB` table is empty, and it will cause a parse error once there's any data in it.

Reply:
Oh, right, I get you now.

Yeah, I think that's along the lines in the other CL of rejecting invalid inputs, which is a lower priority for now.

Reply:
Yeah, that sounds reasonable.